internals/overlord/state/state.go

```diff
@@ -570,6 +568,131 @@ NextChange:
 			delete(s.tasks, tid)
 		}
 	}
+	// Prune expired notices, and update the latest warning time cache.
+	var latestWarningTime time.Time
+	for k, n := range s.notices {
+		if n.expired(now) {
+			stats.IncludeNotice(n)
+			delete(s.notices, k)
+		} else if n.noticeType == WarningNotice {
+			if n.lastRepeated.After(latestWarningTime) {
+				latestWarningTime = n.lastRepeated
+			}
+		} else if n.noticeType == ChangeUpdateNotice {
+			if _, exists := s.changes[n.key]; !exists {
+				stats.IncludeNotice(n)
+				delete(s.notices, k)
+			}
+		}
+	}
+	s.latestWarningTime.Store(&latestWarningTime)
+	if len(s.notices) > maxNotices {
+		s.pruneMaxNotices(maxNotices, stats)
+	}
+	stats.Log()
+}
+
+// pruneStats tracks how many changes and notices have been pruned, and what
+// time range has been affected.
+type pruneStats struct {
+	numChangesPruned   int
+	numNoticesPruned   int
+	oldestChangePruned time.Time
+	newestChangePruned time.Time
+	oldestNoticePruned time.Time
+	newestNoticePruned time.Time
+}
+
+func (p *pruneStats) IncludeChange(chg *Change) {
+	p.numChangesPruned++
+	if chg == nil || chg.readyTime.IsZero() {
+		return
+	}
+	if p.oldestChangePruned.IsZero() || p.oldestChangePruned.After(chg.readyTime) {
+		p.oldestChangePruned = chg.readyTime
+	}
+	if p.newestChangePruned.IsZero() || p.newestChangePruned.Before(chg.readyTime) {
```

Comment:
To me this logic seemed slightly confusing/backward. I think we should say "if *this* change's time is before/after" rather than the other way.

Also, we don't need to check `p.newestChange.IsZero()` because the `readyTime` will always be after zero.

Similar in `IncludeNotice`.

```suggestion
	if p.oldestChange.IsZero() || chg.readyTime.Before(p.oldestChange) {
		p.oldestChange = chg.readyTime
	}
	if chg.readyTime.After(p.newestChange) {
```

Reply:
done